Src/ElasticScale.Client/ShardManagement/Cache/PerfCounterInstance.cs

```diff
@@ -0,0 +1,274 @@
+            }
+            else
+            {
+                // Remove instance, if already exists
+                if (PerformanceCounterCategory.InstanceExists(_instanceName, PerformanceCounters.ShardManagementPerformanceCounterCategory))
+                {
+                    // As performance counters are created with Process lifetime and instance name is unique (PID + shard map name),
+                    // this should never happen. Trace out error and silently continue.
+                    Tracer.TraceWarning(TraceSourceConstants.ComponentNames.PerfCounter,
+                        "create",
+                        "Performance counter instance {0} already exists. no performance data will be collected.", _instanceName);
+                }
+                else
+                {
+                    // now initialize all counters for this instance
+                    _counters = new Dictionary<PerformanceCounterName, PerformanceCounterWrapper>();
+
+                    foreach (PerfCounterCreationData d in PerfCounterInstance.counterList)
+                    {
+                        _counters.Add(d.CounterName,
+                            new PerformanceCounterWrapper(PerformanceCounters.ShardManagementPerformanceCounterCategory, _instanceName, d.CounterDisplayName));
+                    }
+
+                    // check that atleast one performance counter was created, so that we can remove instance as part of Dispose()
+                    _initialized = _counters.Any(c => c.Value._isValid = true);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Try to increment specified performance counter by 1 for current instance.
+        /// </summary>
+        /// <param name="counterName">Counter to increment.</param>
+        internal void IncrementCounter(PerformanceCounterName counterName)
+        {
+            if (_initialized)
+            {
+                PerformanceCounterWrapper pc;
+                if (_counters.TryGetValue(counterName, out pc))
+                {
+                    pc.Increment();
+                }
+            }
+        }
+
+        /// <summary>
+        ///  Try to update performance counter with speficied value.
+        /// </summary>
+        /// <param name="counterName">Counter to update.</param>
+        /// <param name="value">New value.</param>
+        internal void SetCounter(PerformanceCounterName counterName, long value)
+        {
+            if (_initialized)
+            {
+                PerformanceCounterWrapper pc;
+                if (_counters.TryGetValue(counterName, out pc))
+                {
+                    pc.SetRawValue(value);
+                }
+            }
+        }
+        /// <summary>
+        /// Static method to recreate Shard Management performance counter catagory with given counter list.
+        /// </summary>
+        internal static void CreatePerformanceCategoryAndCounters()
```

Comment:
So this has to be run once per machine, or once per process? i.e. does EDCL always need to be run under admin, or just once? If just once I think we should try adding a powershell script that calls into this method and installs the categories. That way the amount of code run under admin is reduced.